internal/rest/leveldbreceipts.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2021 Kaleido
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rest
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/kaleido-io/ethconnect/internal/errors"
+	"github.com/kaleido-io/ethconnect/internal/kvstore"
+)
+
+type levelDBReceipts struct {
+	conf  *LevelDBReceiptStoreConf
+	store kvstore.KVStore
+}
+
+func newLevelDBReceipts(conf *LevelDBReceiptStoreConf) (*levelDBReceipts, error) {
+	store, err := kvstore.NewLDBKeyValueStore(conf.Path)
+	if err != nil {
+		return nil, errors.Errorf(errors.ReceiptStoreLevelDBConnect, err)
+	}
+	return &levelDBReceipts{
+		conf:  conf,
+		store: store,
+	}, nil
+}
+
+// AddReceipt processes an individual reply message, and contains all errors
+// To account for any transitory failures writing to mongoDB, it retries adding receipt with a backoff
+func (l *levelDBReceipts) AddReceipt(requestID string, receipt *map[string]interface{}) (err error) {
+	// insert an entry with a composite key to track the insertion order
+	timeElapsed := ((*receipt)["headers"].(map[string]interface{}))["timeElapsed"].(float64)
+	timeSeconds := (*receipt)["receivedAt"].(int64) + int64(timeElapsed)
+	timeStringSeconds := fmt.Sprintf("%d", timeSeconds)
+	timeStringNanoSeconds := fmt.Sprintf("%.9f", timeElapsed)
+	timeStringNanoSeconds = timeStringNanoSeconds[strings.Index(timeStringNanoSeconds, "."):]
+	timestamp := timeStringSeconds + timeStringNanoSeconds
+
+	// add "z" prefix so these entries come after the actual entries
+	// because for iteration we start from last backwards
+	compositeKey := fmt.Sprintf("z%s:%s", timestamp, requestID)
+	l.store.Put(compositeKey, nil)
```

Comment:
I believe if you swap the logic such that we put the _real_ record under the `compositeKey`, and store the composite key under the `requestID` then you can really simplify the `GetReceipts` logic below.

It would mean that a get-by-ID would need a second Get, which of course is the payoff.

So I'll leave it with you whether you make some of the suggested tweaks to the performance of the `GetReceipts` loop below (particularly avoiding the Regexp compile), or just do the indirected query lookup on the `GetReceipt` (thus eliminating the need for the Regexp or split).

Reply:
Think the GetReceipts() always processing multiple rounds is more performance sensitive, so went with the swap to have the generated key mapped to the real record, and having the requestID mapped to the generated key